core/queueref_test.go

```diff
@@ -56,3 +59,152 @@ func TestQueueRef(t *testing.T) {
+	val := [3]int{10, 20, 30}
+	core.Put("key1", core.NewObj(val[0], -1, core.OBJ_TYPE_STRING, core.OBJ_ENCODING_INT))
+	qr.Insert("key1")
+	core.Put("key2", core.NewObj(val[1], -1, core.OBJ_TYPE_STRING, core.OBJ_ENCODING_INT))
+	qr.Insert("key2")
+	core.Put("key3", core.NewObj(val[2], -1, core.OBJ_TYPE_STRING, core.OBJ_ENCODING_INT))
+	qr.Insert("key3")
+	for i := 0; i < len(val); i++ {
+		if qe, err := qr.Remove(); err != nil || qe.Obj.Value != val[i] {
+			t.Error("removal from queue with multiple non expired keys failed, Expected : ", val[i], " Got : ", qe.Obj.Value)
+			break
+		}
+	}
+}
+
+// Test for removing from queue with expired keys before non-expired
+func TestRemoveExpiredBeforeNonExpire(t *testing.T) {
+	qr := core.NewQueueRef()
+	val := [3]int{10, 20}
+	core.Put("key1", core.NewObj(val[0], 1, core.OBJ_TYPE_STRING, core.OBJ_ENCODING_INT))
+	qr.Insert("key1")
+	core.Put("key2", core.NewObj(val[1], -1, core.OBJ_TYPE_STRING, core.OBJ_ENCODING_INT))
+	qr.Insert("key2")
+	time.Sleep(2 * time.Millisecond)
+	if qe, err := qr.Remove(); err != nil || qe.Obj.Value != val[1] {
+		t.Error("test for removing expired key before non-expired failed , Expected : ", val[1], " Got : ", qe.Obj.Value)
+	}
+}
+
+// Test for removing from queue with multiple expired keys before non-expired
+func TestRemoveMultipleExpiredBeforeNonExpire(t *testing.T) {
+	qr := core.NewQueueRef()
+	val := [3]int{10, 20, 30}
+	core.Put("key1", core.NewObj(val[0], 1, core.OBJ_TYPE_STRING, core.OBJ_ENCODING_INT))
+	qr.Insert("key1")
+	core.Put("key2", core.NewObj(val[1], 1, core.OBJ_TYPE_STRING, core.OBJ_ENCODING_INT))
+	qr.Insert("key2")
+	core.Put("key3", core.NewObj(val[2], -1, core.OBJ_TYPE_STRING, core.OBJ_ENCODING_INT))
+	qr.Insert("key3")
+	time.Sleep(2 * time.Millisecond)
+	if qe, err := qr.Remove(); err != nil || qe.Obj.Value != val[2] {
+		t.Error("test for removing mulitple expired key before non-expired failed , Expected : ", val[2], " Got : ", qe.Obj.Value)
+	}
+}
+
+// Test for removing from queue with all expired keys
+func TestRemoveAllExpired(t *testing.T) {
+	qr := core.NewQueueRef()
+	val := [3]int{10, 20, 30}
+	core.Put("key1", core.NewObj(val[0], 1, core.OBJ_TYPE_STRING, core.OBJ_ENCODING_INT))
+	qr.Insert("key1")
+	core.Put("key2", core.NewObj(val[1], 1, core.OBJ_TYPE_STRING, core.OBJ_ENCODING_INT))
+	qr.Insert("key2")
+	core.Put("key3", core.NewObj(val[2], 1, core.OBJ_TYPE_STRING, core.OBJ_ENCODING_INT))
+	qr.Insert("key3")
+	time.Sleep(2 * time.Millisecond)
+	if _, err := qr.Remove(); err == nil {
+		t.Error("test for removing all expired key failed, Expected : ", core.ErrQueueEmpty, " Got : ", err)
+	}
+	// remove from empty queue
+	if _, err := qr.Remove(); err == nil {
+		fmt.Println(err)
+		t.Error("test for removing from empty queue failed Expected : ", core.ErrQueueEmpty, " Got : ", err)
+	}
+}
+
+// Benchmark queueref by inserting expired, non-expired and expired keys in order and removing them
+func BenchmarkQueueRef(b *testing.B) {
```

Comment:
I have a few suggestions to improve the benchmark:

1. Let's split the benchmark into separate operations - insertion and removal. This will give us more granular performance insights for each operation.
2. Please use `b.Run()` calls to create sub-benchmarks for different data sizes and operations.
3. Call `b.ResetTimer()` to ensure we're only timing the operations we intend to measure, not the setup code. Once the setup is complete (initializing queue, dataset, channel, etc.) make a call to `b.ResetTimer()` to ensure these operations don't affect the benchmark results.

You can probably structure it this way:

```go
func BenchmarkQueueRef(b *testing.B) {
    benchmarkCases := []struct {
        name         string
        expiredFirst int
        nonExpired   int
        expiredLast  int
    }{
        {"Small", 10, 20, 30},
        {"Medium", 100, 200, 300},
        {"Large", 100, 2000, 3000},
        {"VeryLarge", 10000, 20000, 30000},
    }

    for _, bc := range benchmarkCases {
        // Setup code...

        b.Run(fmt.Sprintf("%s_Insert", bc.name), func(b *testing.B) {
            for i := 0; i < b.N; i++ {
                // Insertion benchmark...
            }
        })

        b.Run(fmt.Sprintf("%s_Remove", bc.name), func(b *testing.B) {
            for i := 0; i < b.N; i++ {
                // Removal benchmark...
            }
        })
    }
}

Reply:
Thank you for the suggestion, I will update.

Reply:
what's the best way to benchmark removal, since during benchmark the queue might get emptied? 
So should we stop the loop when queue gets emptied?

Reply:
I have updated the benchmark, please review it